google/cloud/storage/status_or.h

```diff
@@ -290,13 +295,6 @@ class StatusOr final {
     }
   }
 
-  void reset() {
-    if (ok()) {
-      reinterpret_cast<T*>(&buffer_)->~T();
-      status_ = Status(StatusCode::UNKNOWN, std::string{});
-    }
-  }
-
   Status status_;
   using aligned_storage_t = std::aligned_storage<sizeof(T), alignof(T)>;
```

Comment:
Are we ">= C++11"?

If so, is there an alternative here that avoids aligned_storage and all the reinterpret_casts that looks like ...

`union { T value_; }
`

?

Reply:
Yes, we are >= C++11.  Tell me more about this trick.  Are you thinking of the stuff used for optional in Abseil:

https://github.com/abseil/abseil-cpp/blob/7ffbe09f3d85504bd018783bbe1e2c12992fe47c/absl/types/optional.h#L166

Or is it something else?  If it is the Abseil trick, that is cool, but I believe it still calls the destructor explicitly (no `reinterpret_cast<>` though).

Reply:
Yes, something like that.  Yes, you still need the placement news and the explicit destructors, but they become simply ...

```
new (&value_) T(...);
value_.~T();
```
..., and the casts in the accessors disappear too.

Of course, I haven't actually tried any of this.  :blush:

Reply:
Worth a shot, I made the changes, tested with GCC and Clang on my workstation, let's see what the other compilers think.

Reply:
Am I correct in believing that everything looks good?

Reply:
Yup, the compilers are all happy.